### PR TITLE
Upgrade WebRender

### PR DESCRIPTION
**Pull request by pcwalton**
This is against an old Servo because I can't build current Servo due to compile errors in RNG crates. I verified that it starts up.

---

- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes fix #23516
- [x] There are tests for these changes

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/23516)

### CODE REVIEW

**Comment by jdm** [comment]
@pcwalton The RNG crate problems can usually be resolved by upgrading the crates in question.

**Comment by pcwalton** [comment]
OK, this one should work.

**Comment by jdm** [comment]
This should have addressed all of the previous test failures. Thank you for doing all of the preliminary work, @pcwalton!

**Comment by Eijebong** [comment]
```
error: unused variable: `clip`
    --> components/layout/display_list/builder.rs:1446:9
     |
1446 |         clip: Rect<Au>,
     |         ^^^^ help: consider prefixing with an underscore: `_clip`
     |
     = note: `-D unused-variables` implied by `-D warnings`

error: aborting due to previous error
```

Also this needs to be rebased after the diverse rand update thingy (it should dedupe ws and sha1 too)

**Comment by jdm** [comment]
I'm attempting to determine why the fuzzy annotation that I added to allow the test to pass locally is not causing the test to pass on linux CI. @pcwalton in the meantime could you review the extra commits that I added?

**Comment by jdm** [comment]
Huh, I totally misinterpreted the log. The test run is timing out, and I can reproduce this when running the single test on a Linux CI machine. That's interesting.

**Comment by jdm** [comment]
It's very strange - I can reproduce the timing out behaviour consistently on the Servo Debian CI machines. I can't reproduce it on a personal Ubuntu machine.

**Comment by jdm** [comment]
Ok, so based on the output of [py-spy](https://crates.io/crates/py-spy), the reason for the timeout is that there's a deadlock on the lock inside mozlog's structured logger. This is being tickled by:
```
Thread 0x7F8622FFD700 (active)
         _handle_log (/home/servo/jdm/servo/python/_virtualenv/lib/python2.7/site-packages/mozlog/structuredlog.py:260)
         _log_data (/home/servo/jdm/servo/python/_virtualenv/lib/python2.7/site-packages/mozlog/structuredlog.py:254)
         log (/home/servo/jdm/servo/python/_virtualenv/lib/python2.7/site-packages/mozlog/structuredlog.py:551)
         inner (/home/servo/jdm/servo/python/_virtualenv/lib/python2.7/site-packages/mozlog/logtypes.py:55)
         emit (/home/servo/jdm/servo/python/_virtualenv/lib/python2.7/site-packages/mozlog/stdadapter.py:19)
         handle (/home/servo/jdm/servo/python/_virtualenv/lib/python2.7/site-packages/mozlog/stdadapter.py:22)
         callHandlers (logging/__init__.py:1336)
         handle (logging/__init__.py:1296)
         _log (logging/__init__.py:1286)
         debug (logging/__init__.py:1155)
         call (/home/servo/jdm/servo/python/_virtualenv/lib/python2.7/site-packages/PIL/PngImagePlugin.py:146)
         _open (/home/servo/jdm/servo/python/_virtualenv/lib/python2.7/site-packages/PIL/PngImagePlugin.py:582)
         __init__ (/home/servo/jdm/servo/python/_virtualenv/lib/python2.7/site-packages/PIL/ImageFile.py:103)
         _open_core (/home/servo/jdm/servo/python/_virtualenv/lib/python2.7/site-packages/PIL/Image.py:2676)
         open (/home/servo/jdm/servo/python/_virtualenv/lib/python2.7/site-packages/PIL/Image.py:2690)
         get_differences (/home/servo/jdm/servo/tests/wpt/web-platform-tests/tools/wptrunner/wptrunner/executors/base.py:321)
         is_pass (/home/servo/jdm/servo/tests/wpt/web-platform-tests/tools/wptrunner/wptrunner/executors/base.py:307)
         run_test (/home/servo/jdm/servo/tests/wpt/web-platform-tests/tools/wptrunner/wptrunner/executors/base.py:356)
         do_test (/home/servo/jdm/servo/tests/wpt/web-platform-tests/tools/wptrunner/wptrunner/executors/executorservo.py:267)
         run_test (/home/servo/jdm/servo/tests/wpt/web-platform-tests/tools/wptrunner/wptrunner/executors/base.py:192)
         run_test (/home/servo/jdm/servo/tests/wpt/web-platform-tests/tools/wptrunner/wptrunner/testrunner.py:115)
         run (/home/servo/jdm/servo/tests/wpt/web-platform-tests/tools/wptrunner/wptrunner/testrunner.py:98)
         start_runner (/home/servo/jdm/servo/tests/wpt/web-platform-tests/tools/wptrunner/wptrunner/testrunner.py:149)
         run (multiprocessing/process.py:114)
         _bootstrap (multiprocessing/process.py:258)
         __init__ (multiprocessing/forking.py:126)
         start (multiprocessing/process.py:130)
         start_test_runner (/home/servo/jdm/servo/tests/wpt/web-platform-tests/tools/wptrunner/wptrunner/testrunner.py:499)
         init (/home/servo/jdm/servo/tests/wpt/web-platform-tests/tools/wptrunner/wptrunner/testrunner.py:476)
         run (/home/servo/jdm/servo/tests/wpt/web-platform-tests/tools/wptrunner/wptrunner/testrunner.py:360)
         __bootstrap_inner (threading.py:801)
         __bootstrap (threading.py:774)
```
and there are a bunch of other threads waiting on it as well.

**Comment by jdm** [comment]
Submitted https://github.com/web-platform-tests/wpt/pull/17564 upstream to fix the deadlock. Once that merges and we sync up, we'll be able to finally merge this work!

@pcwalton Please review the extra commits that I added on top of your work!

File: components/layout/sequential.rs [hunk 1]
```diff
@@ -72,11 +73,31 @@ pub fn reflow(root: &mut dyn Flow, layout_context: &LayoutContext, relayout_mode
...
     flow_root.collect_stacking_contexts(&mut state);
 
-    let state = DisplayListBuildState::new(layout_context, state);
+    let mut state = DisplayListBuildState::new(layout_context, state);
+
+    // Create a base rectangle for the page background based on the root
+    // background color.
+    let base = state.create_base_display_item(
+        Rect::new(Point2D::new(Au::new(0), Au::new(0)), client_size),
+        Rect::new(Point2D::new(Au::new(0), Au::new(0)), client_size),
```
Comment (pcwalton) [inline components/layout/sequential.rs]: I think there's a `ZERO_POINT` constant somewhere

File: components/layout/display_list/builder.rs [hunk 2]
```diff
@@ -1464,7 +1464,7 @@ impl Fragment {
             let background_color = style.resolve_color(style.get_background().background_color);
             let base = state.create_base_display_item(
                 stacking_relative_border_box,
-                clip,
+                stacking_relative_border_box,
```
Comment (pcwalton) [inline components/layout/display_list/builder.rs]: Should the title of this commit be "WebRender rectangle items are *not* sized based on their clipping rectangles"? I'm not sure what the commit message means here, if you're removing `clip`.

Comment (jdm) [reply components/layout/display_list/builder.rs]: The second argument is the clipping rectangle, which controls the size of the rectangle.

Comment (pcwalton) [reply components/layout/display_list/builder.rs]: OK, I guess that's fine.

**Comment by pcwalton** [comment]
Looks good to me.

**Comment by jdm** [comment]
Once https://github.com/servo/servo/pull/23732 merges we will need to upstream https://github.com/servo/webrender/compare/servo-hl and then upgrade webrender again so as not to regress the HoloLens build.

**Comment by atouchet** [comment]
Could dwrote be updated in gfx as part of this to get rid of the duplicate crate?

**Comment by CYBAI** [comment]
In `linux`, `android`, `magicleap` and `windows` build:

```rust
error[E0432]: unresolved imports `servo::webrender_api::DevicePixel`, `servo::webrender_api::FramebufferPixel`
  --> ports/libsimpleservo/api/src/lib.rs:25:28
   |
25 | use servo::webrender_api::{DevicePixel, FramebufferPixel, ScrollLocation};
   |                            ^^^^^^^^^^^  ^^^^^^^^^^^^^^^^ no `FramebufferPixel` in `webrender_api`
   |                            |
   |                            no `DevicePixel` in `webrender_api`

error: aborting due to previous error

For more information about this error, try `rustc --explain E0432`.
error: Could not compile `simpleservo`.
```

**Comment by CYBAI** [comment]
This one pass again in the second Mac build 👀 other failures in other mac builds are #23290 

```
{
  "status": "PASS",
  "group": "default",
  "message": null,
  "stack": null,
  "subtest": null,
  "test": "/css/compositing/mix-blend-mode/mix-blend-mode-both-parent-and-blended-with-3D-transform.html",
  "line": 2057,
  "action": "test_result",
  "expected": "FAIL"
}
```

besides, in `magicleap` build:

```rust
error[E0432]: unresolved import `servo::webrender_api::DevicePixel`
  --> ports/libmlservo/src/lib.rs:17:5
   |
17 | use servo::webrender_api::DevicePixel;
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ no `DevicePixel` in `webrender_api`

error[E0432]: unresolved import `servo::webrender_api::DevicePoint`
  --> ports/libmlservo/src/lib.rs:18:5
   |
18 | use servo::webrender_api::DevicePoint;
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ no `DevicePoint` in `webrender_api`

error[E0432]: unresolved import `servo::webrender_api::LayoutPixel`
  --> ports/libmlservo/src/lib.rs:19:5
   |
19 | use servo::webrender_api::LayoutPixel;
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ no `LayoutPixel` in `webrender_api`

error: aborting due to 3 previous errors
```

**Pull request merged**